### ISSUE TITLE
Getting Started CHE CLI: che.sh fails to download cli.sh

### ISSUE THREAD

**Issue by elonmallin** [eclipse-che/che#3266]
che.sh seems to have a problem with the command `docker run --rm --net=host appropriate/curl "$@"` which is evaluated when you run `che start` like in the *reproduction steps*. I tried that command manually without all the curl silent flags `docker run --rm --net=host appropriate/curl --head --fail https://raw.githubusercontent.com/eclipse/che/master/cli.sh` to get more information, and the output was:
```
curl: (6) Couldn't resolve host 'raw.githubusercontent.com'
```
I ran it multiple times with the same result, meanwhile I was able to access `raw.githubusercontent.com...` from my PC and running my locally installed curl succeeds so the problem seems to be running curl from this docker container.

**Reproduction Steps:**

1. Follow the[ Getting started guide](https://eclipse-che.readme.io/docs/che-getting-started#1-get-the-eclipse-che-cli) for windows.
Error appears when you run `che start`.

**Observed behavior:**

Output: `CLI download error. Bad network or version. CLI update works for 5.0.0-M3+.`

**OS and version:** Window 10 Pro
**Docker version:** 1.12.3 (The native windows version)

**Comment by elonmallin** [comment]
One possible solution would be not to use the docker container to run curl at all since the getting started guide already asks you to download che.sh with curl it's fair to assume that the user has it. The better solution would be to make this work and also use the docker container in the getting started guide to be consistent and make it easier for windows users that don't have curl by default.

**Comment by TylerJewell** [comment]
Hi @elonmallin - that particular error that you are getting would seem to indicate that you are behind a firewall. In the current version of the CLI, the remainder of the CLI is downloaded directly from GitHub using curl.  We have a new version of the CLI that we will ship in a week or two that creates a cleaner abstraction and this issue would not exist.

Are you - by chance - behind a proxy?  If so, then for Docker to work behind a proxy that way, there may be some additional environment variables that you need to pass into the run command.

**Comment by elonmallin** [comment]
Hi @TylerJewell okey I tried deactivating my windows firewall but that doesn't do anything, and I don't have any extra installed software for firewalls so I don't know. But doesn't most modems just have firewalls in them too? I'm testing at home and I'm not behind any proxy unless there is some hidden one by my ISP. But my locally installed version of curl also works so doesn't that indicate that the docker container doesn't have the same access rights to the network or something? But either way I guess it won't matter if you are releasing a new version that doesn't use this docker curl image.

**Comment by elonmallin** [comment]
I tried commenting out the `curl` function in the che.sh so that I could do the the `che start` command and that worked. After that I wanted to try the `che dir init` command (in a different folder after I stopped the running che server) to get a chefile with a bunch of template stuff in there (I read somewhere) but that command does nothing. I only get `WARNING: No kernel memory limit support` before exit.

So I checked the [docs](https://eclipse-che.readme.io/docs/installation#from-source) again and found a docker command to run that you can apparently do instead of the `che dir` command but it is using an image that doesn't exist anymore `eclipse/che-file`. Created an issue for that #3269.

**Comment by TylerJewell** [comment]
That looks like a typo in the docs - it should be eclipse/che-dir, not che-file. I updated the docs.

As for why the curl command is not working - if you commented out the curl command, but other commands work, then chances are you are not behind a proxy, as there is other software that would fail if you were behind the proxy.

I am curious if you have the same issue on the new CLI.  The syntax is hidden, but you can view it here - and it should be workable for testing.

https://github.com/eclipse/che/tree/master/docs

**Comment by elonmallin** [comment]
I'm not sure what to run anymore exactly. The `dir` command is not mentioned anymore but looks like it could have been abstracted away or something. Also not sure what the `${CHE_MINI_PRODUCT_NAME}` should be that is mentioned in the docs but I guess `eclipse`. Anyway, I ran: `docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock -v /d/Projects/Tests/che-dir:/che eclipse/che-cli:nightly init` and it created these files:
```
+---config
|   |   che.env
|   |   che.ver
|   |   
|   \---docs
|           DOCS.md
|           README.md
|           
\---instance
```
So no Chefile with some template stuff in there like I thought. But I could also look at the docs and just create a Chefile and look at the examples in there.
`docker ...che-cli start` seems to fail with this:
```
INFO: (che cli):Loading cli...
INFO: (che cli):Checking registry for version 'nightly' images
INFO: (che download): Checking for image 'puppet/puppet-agent-alpine:4.6.2'...found
INFO: (che download): Checking for image 'registry:2.5.0'...found
INFO: (che download): Checking for image 'eclipse/che-init:nightly'...found
INFO: (che download): Checking for image 'eclipse/che-server:nightly'...found
INFO: (che init):Installing configuration and bootstrap variables:
INFO: (che init):  CHE_HOST=10.0.75.2
INFO: (che init):  CHE_PORT=8080
INFO: (che init):  CHE_VERSION=nightly
INFO: (che init):  CHE_CONFIG=/d/Projects/Tests/che-dir/config
INFO: (che init):  CHE_INSTANCE=/d/Projects/Tests/che-dir/instance
INFO: (che init):  CHE_ENVIRONMENT=production
INFO: (che config):CHE_VERSION=nightly does not match che.env=<not-installed>. Aborting.
INFO: (che config):This happens if the <version> of your Docker image is different from /d/Projects/Tests/che-dir/config/che.env
```
And there is no mention of the word "version" in che.env so donno.

**Comment by elonmallin** [comment]
Just fyi what I wanted to try kind of was to use the cli to init and empty directory and get a chefile and then point it to like a nodejs image and get a small webserver running and just try out the flow of creating a new workspace with che.

**Comment by TylerJewell** [comment]
Something is definitely not right.

First, the che-dir utility and the che-cli are separate concepts. They are not really intended to be mixed and matched.  So it seems that you are trying to mix things together.

Second, I don't think you are running the right version of the software.  Delete the directory that you created for Che, then do a "docker pull eclipse/che-cli:nightly", and then rerun the initial start command in an empty directory. This should work properly the first go around.

**Comment by elonmallin** [comment]
Okey see now there is a `eclipse/che-dir` and `eclipse/che-cli` and they are different. They merged in my mind because of the starter docs I first referenced. Since it has you download the `che.sh` and I went into the code of that and saw it downloads `cli.sh` from github and that cli does have a `dir` command as opposed to `eclipse/che-cli`. Too many command line tools to keep track of but as a user I'm just supposed to know about the `che.sh` so me debugging and finding all these separate things caused some confusion.

Anyway, created a new folder somewhere else and did the `pull` to update and then the `start` command but all I got now was just one line then exit.
```
INFO: (che cli): Loading cli...
```

But I already successfully got che running with the `che.sh` from the getting started docs with the `start` command (when I commented out the curl function to just use my locally installed curl).

But maybe we should leave this issue at the network error I got from the cli using the `appropriate/curl` command should be fixed in a later update as you mentioned and my workaround was to comment out the `curl` command in the `che.sh`.

The "issue" I got now, running what you asked, is related to something else I guess and is probably just under development so no other users should see this error atm unless they are just testing the current nightly build I guess. I joined the gitter channel so we could carry on there if you want to keep test with the nightly. But I think the original issue can be closed or should be closed once the new release comes out and we can keep this issue a bit more useful for anyone coming to read it later?

**Comment by TylerJewell** [comment]
Thanks for explaining.  I am thinking that the appropriate/curl issue that you had before is also affecting this new CLI.  I need to get hands on with your particular system to diagnose it - so we would need to schedule a screen share. I am not sure you can give us any more diagnostic information.

One other thing that you can try is to bypass the CLI alltogether. This just runs hte che server without any fancy CLI stuff in the middle. There is no curl image on this one.

https://eclipse-che.readme.io/v5.0/docs/usage-docker-server

**Comment by elonmallin** [comment]
Okey I see, yea we can do that. I'm on UTC+1 though so would have to be sometime in the evening for me and day for you I guess. I'm in the eclipse/che gitter or you can reach me through my profile here to set something up.

I can get it running with the che.sh when I use my local curl though so don't think I need to run the docker image directly but that is basically all the cli does if you just do `start` or?

**Comment by TylerJewell** [comment]
You should still try launching with the usage-docker-server link before we get together for a debug session.  yes, the CLI container uses curl, but it's doing a lot of environmental setup to (hopefully) make the boot of Che a lot easier.  If you can run the usage-docker-server approach cleanly, then that tells me that there is something about the docker configuration that I need to inspect on your system.

**Comment by elonmallin** [comment]
Ah okey. Yea I was able to run that and go to localhost:8080, got this output:
```
Using embedded assembly...
Using embedded che.properties... Copying template to /d/Projects/Tests/che-dir/conf.
2016-12-05 08:15:41,155[main]             [INFO ] [o.a.c.s.VersionLoggerListener 89]    - Server version:        Apache Tomcat/8.0.38
2016-12-05 08:15:41,157[main]             [INFO ] [o.a.c.s.VersionLoggerListener 91]    - Server built:          Oct 6 2016 20:51:55 UTC
2016-12-05 08:15:41,157[main]             [INFO ] [o.a.c.s.VersionLoggerListener 93]    - Server number:         8.0.38.0
2016-12-05 08:15:41,158[main]             [INFO ] [o.a.c.s.VersionLoggerListener 95]    - OS Name:               Linux
2016-12-05 08:15:41,158[main]             [INFO ] [o.a.c.s.VersionLoggerListener 97]    - OS Version:            4.4.27-moby
2016-12-05 08:15:41,159[main]             [INFO ] [o.a.c.s.VersionLoggerListener 99]    - Architecture:          amd64
2016-12-05 08:15:41,159[main]             [INFO ] [o.a.c.s.VersionLoggerListener 101]   - Java Home:             /usr/lib/jvm/java-1.8-openjdk/jre
2016-12-05 08:15:41,159[main]             [INFO ] [o.a.c.s.VersionLoggerListener 103]   - JVM Version:           1.8.0_101-b13
2016-12-05 08:15:41,160[main]             [INFO ] [o.a.c.s.VersionLoggerListener 105]   - JVM Vendor:            Oracle Corporation
2016-12-05 08:15:41,160[main]             [INFO ] [o.a.c.s.VersionLoggerListener 107]   - CATALINA_BASE:         /home/user/che/tomcat
2016-12-05 08:15:41,160[main]             [INFO ] [o.a.c.s.VersionLoggerListener 109]   - CATALINA_HOME:         /home/user/che/tomcat
2016-12-05 08:15:41,161[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Djava.util.logging.config.file=/home/user/che/tomcat/conf/logging.properties
2016-12-05 08:15:41,161[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Djava.util.logging.manager=org.apache.juli.ClassLoaderLogManager
2016-12-05 08:15:41,162[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dche.docker.network=bridge
2016-12-05 08:15:41,162[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dport.http=8080
2016-12-05 08:15:41,162[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dche.home=/home/user/che
2016-12-05 08:15:41,162[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dche.logs.dir=/home/user/che/tomcat/logs/
2016-12-05 08:15:41,163[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dche.logs.level=INFO
2016-12-05 08:15:41,163[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Djdk.tls.ephemeralDHKeySize=2048
2016-12-05 08:15:41,163[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Djava.protocol.handler.pkgs=org.apache.catalina.webresources
2016-12-05 08:15:41,163[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dcom.sun.management.jmxremote
2016-12-05 08:15:41,164[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dcom.sun.management.jmxremote.ssl=false
2016-12-05 08:15:41,164[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dcom.sun.management.jmxremote.authenticate=false
2016-12-05 08:15:41,164[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dche.local.conf.dir=/data/conf
2016-12-05 08:15:41,165[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Djava.endorsed.dirs=/home/user/che/tomcat/endorsed
2016-12-05 08:15:41,165[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dcatalina.base=/home/user/che/tomcat
2016-12-05 08:15:41,165[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Dcatalina.home=/home/user/che/tomcat
2016-12-05 08:15:41,166[main]             [INFO ] [o.a.c.s.VersionLoggerListener 115]   - Command line argument: -Djava.io.tmpdir=/home/user/che/tomcat/temp
2016-12-05 08:15:41,249[main]             [INFO ] [o.a.c.http11.Http11NioProtocol 443]  - Initializing ProtocolHandler ["http-nio-8080"]
2016-12-05 08:15:41,265[main]             [INFO ] [o.a.t.util.net.NioSelectorPool 72]   - Using a shared selector for servlet write/read
2016-12-05 08:15:41,268[main]             [INFO ] [o.a.catalina.startup.Catalina 591]   - Initialization processed in 376 ms
2016-12-05 08:15:41,299[main]             [INFO ] [c.m.JmxRemoteLifecycleListener 332]  - The JMX Remote Listener has configured the registry on port 32001 and the server on port 32101 for the Platform server
2016-12-05 08:15:41,369[main]             [INFO ] [o.a.c.core.StandardService 435]      - Starting service Catalina
2016-12-05 08:15:41,369[main]             [INFO ] [o.a.c.core.StandardEngine 259]       - Starting Servlet Engine: Apache Tomcat/8.0.38
2016-12-05 08:15:41,469[ost-startStop-1]  [INFO ] [o.a.c.startup.HostConfig 911]        - Deploying web application archive /home/user/che/tomcat/webapps/ROOT.war
2016-12-05 08:15:42,721[ost-startStop-1]  [INFO ] [o.a.c.startup.HostConfig 975]        - Deployment of web application archive /home/user/che/tomcat/webapps/ROOT.war has finished in 1,251 ms
2016-12-05 08:15:42,722[ost-startStop-1]  [INFO ] [o.a.c.startup.HostConfig 911]        - Deploying web application archive /home/user/che/tomcat/webapps/swagger.war
2016-12-05 08:15:42,793[ost-startStop-1]  [INFO ] [o.a.c.startup.HostConfig 975]        - Deployment of web application archive /home/user/che/tomcat/webapps/swagger.war has finished in 71 ms
2016-12-05 08:15:42,794[ost-startStop-1]  [INFO ] [o.a.c.startup.HostConfig 911]        - Deploying web application archive /home/user/che/tomcat/webapps/dashboard.war
2016-12-05 08:15:42,962[ost-startStop-1]  [INFO ] [o.a.c.startup.HostConfig 975]        - Deployment of web application archive /home/user/che/tomcat/webapps/dashboard.war has finished in 168 ms
2016-12-05 08:15:42,963[ost-startStop-1]  [INFO ] [o.a.c.startup.HostConfig 911]        - Deploying web application archive /home/user/che/tomcat/webapps/wsmaster.war
2016-12-05 08:15:53,851[ost-startStop-1]  [INFO ] [o.a.c.startup.HostConfig 975]        - Deployment of web application archive /home/user/che/tomcat/webapps/wsmaster.war has finished in 10,888 ms
2016-12-05 08:15:53,863[main]             [INFO ] [o.a.c.http11.Http11NioProtocol 488]  - Starting ProtocolHandler ["http-nio-8080"]
2016-12-05 08:15:53,869[main]             [INFO ] [o.a.catalina.startup.Catalina 642]   - Server startup in 12599 ms
2016-12-05 09:15:53,575[ted-scheduler-0]  [WARN ] [p.d.m.c.DockerContainerCleaner 113]  - Unused container with 'id': 'ac34a88c48d6422c677f6bfa1b7eb443b1d5aace9ed3cc1a846381deab4a4a0a' and 'name': '/workspaceexdgsd45ym56199f_machinerrr1k4glbgk6d7c5_che_dev-machine' was removed
2016-12-05 09:15:53,577[ted-scheduler-0]  [INFO ] [p.d.m.c.DockerContainerCleaner 88]   - List containers registered in the api: []
2016-12-05 10:15:53,495[ted-scheduler-0]  [INFO ] [p.d.m.c.DockerContainerCleaner 88]   - List containers registered in the api: []
2016-12-05 11:15:53,491[ted-scheduler-3]  [INFO ] [p.d.m.c.DockerContainerCleaner 88]   - List containers registered in the api: []
```

**Comment by TylerJewell** [comment]
Yep, then the issue is related to how curl is accessing the internet from inside a docker container on your system.  I would think that you could run our core CLI by adding -e "http_proxy=<proxy>" and -e "no_proxy=<no_proxy>".  There is definitely something on your system that is preventing curl from running properly.  

Given that the direct way worked, we have isolated the issue down to that part.  But since you can run the che-server directly and that seems to be ok.

**Comment by elonmallin** [comment]
Okey, not sure what you mean that I should run though, which one is the core CLI? I tried commenting the curl function back in in the che.sh and then tried to run that with the `-e` flags and I also tried the `eclipse/che-cli:nightly` with the flags. The flags look really weird to me and <no_proxy> just looks like I should replace it with something or. This is the output from a few tries:
```
D:\Projects\Tests\che-cli>che start -e "http_proxy=" -e "no_proxy=<no_proxy>"
WARNING: No kernel memory limit support
ERROR: che start/stop/start: You passed unknown options.

D:\Projects\Tests\che-cli>che start -e "http_proxy="
WARNING: No kernel memory limit support
ERROR: che start/stop/start: You passed unknown options.

D:\Projects\Tests\che-cli>che start -e "no_proxy=<no_proxy>"
WARNING: No kernel memory limit support
ERROR: che start/stop/start: You passed unknown options.

D:\Projects\Tests\che-cli>docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock -v /d/Projects/Tests/che-cli:/che eclipse/che-cli:nightly start -e "http_proxy="
INFO: (che cli): Loading cli...

D:\Projects\Tests\che-cli>docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock -v /d/Projects/Tests/che-cli:/che eclipse/che-cli:nightly start -e "http_proxy=" -e "no_proxy=<no_proxy>"
INFO: (che cli): Loading cli...
```

**Comment by TylerJewell** [comment]
You have to provide your system's PROXY or NO_PROXY URL as values to those parameters.  You keep telling us that curl is not working and the only way that it's not working is if you have a proxy or firewall issue :)

**Comment by elonmallin** [comment]
I'm pretty sure I'm not behind any proxy. Checked my windows settings again and my proxy settings are set to off and previously checked on http://www.lagado.com/proxy-test (plus the cache test) and http://amibehindaproxy.com/.

I tried disabling my firewall in windows as I tested earlier and I don't have any other software that should have any firewall. The only thing could be in the router which I logged in to now and I don't see any options about firewall at all.

Don't know about this stuff but I haven't configured anything special. I don't know what a NO_PROXY URL is so maybe that's something I can still do something about and pass in as arg.

Yes curl through the docker container appropriate/curl is not working but the locally installed one is.
```
C:\cmdstart>docker run --rm --net=host appropriate/curl --head https://raw.githubusercontent.com/eclipse/che/master/cli.sh
curl: (6) Couldn't resolve host 'raw.githubusercontent.com'

C:\cmdstart>curl --head https://raw.githubusercontent.com/eclipse/che/master/cli.sh
HTTP/1.1 200 OK
Content-Security-Policy: default-src 'none'; style-src 'unsafe-inline'
...
```

Checked my docker settings too and "Proxies" there is empty and "Network" only has the defaults set.

**Comment by TylerJewell** [comment]
@riuvshin - any theories on what might cause a container using appropriate/curl to fail on a lookup when it does not on the host?

```
C:\cmdstart>docker run --rm --net=host appropriate/curl --head https://raw.githubusercontent.com/eclipse/che/master/cli.sh
curl: (6) Couldn't resolve host 'raw.githubusercontent.com'

C:\cmdstart>curl --head https://raw.githubusercontent.com/eclipse/che/master/cli.sh
HTTP/1.1 200 OK
Content-Security-Policy: default-src 'none'; style-src 'unsafe-inline'
```

**Comment by riuvshin** [comment]
@elonmallin Hi Elon! just curious can you try to curl any other host from a container? like google.com, I think it may be related to DNS resolver on your host. you can try to specify DNS resolver for particular container, here is docs https://docs.docker.com/engine/userguide/networking/default_network/configure-dns/

**Comment by elonmallin** [comment]
Hi @riuvshin - yea, I think you might be onto something. I got the IP of google and just entered that directly and that seems to work. Using google.com gives an error though just like before however this time number 7 instead of 6.
```
D:\Projects\Tests\che-cli>docker run --rm --net=host appropriate/curl --head google.com
curl: (7) Couldn't connect to server

D:\Projects\Tests\che-cli>curl --head google.com
HTTP/1.1 302 Found
Cache-Control: private
Content-Type: text/html; charset=UTF-8
Location: http://www.google.de/?gfe_rd=cr&ei=jPVHWMHaO7Db8Ael2Z_YDQ
Content-Length: 258
Date: Wed, 07 Dec 2016 11:42:04 GMT

D:\Projects\Tests\che-cli>docker run --rm --net=host appropriate/curl --head 216.58.210.14
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0HTTP/1.1 301 Moved Permanently
Location: http://www.google.com/
Content-Type: text/html; charset=UTF-8
Date: Wed, 07 Dec 2016 11:57:14 GMT
Expires: Fri, 06 Jan 2017 11:57:14 GMT
Cache-Control: public, max-age=2592000
Server: gws
Content-Length: 219
X-XSS-Protection: 1; mode=block
X-Frame-Options: SAMEORIGIN

  0   219    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
```

Looked at the docs and did some more testing and it seems my DNS is my router? I don't know about this stuff so just trying things but I used the `--dns` flag and that seems to have worked:
```
D:\Projects\Tests\che-cli>nslookup google.com
Server:  o2.box
Address:  192.168.1.1

Non-authoritative answer:
Name:    google.com
Addresses:  2a00:1450:4001:81b::200e
          172.217.22.46

D:\Projects\Tests\che-cli>docker run --rm --net=host --dns=192.168.1.1 appropriate/curl --head google.com
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0HTTP/1.1 302 Found
  0   258    0     0    0     0      0      0 --:-Cache-Control: private
Content-Type: text/html; charset=UTF-8
Location: http://www.google.de/?gfe_rd=cr&ei=k_xHWM_6Fqmg8wfUmqCwAQ
Content-Length: 258
Date: Wed, 07 Dec 2016 12:12:03 GMT

D:\Projects\Tests\che-cli>docker run --rm --net=host appropriate/curl --head google.com
curl: (6) Couldn't resolve host 'google.com'
```
If you have any idea what my router is doing if this is strange behavior or what, Or if I could fix this by some setting in docker I would appreciate it. I also went in to the docker settings UI on windows in the "Network" tab and changed `DNS Server` from `automatic` to `fixed 192.168.1.1` but that didn't help.

**Comment by ghost** [comment]
Hi,

Maybe playing with a DNS server will help? Can you try setting it to 8.8.8.8 and not automatic - I bet it's automatic on your Win machine.

![image](https://cloud.githubusercontent.com/assets/5337267/20968767/aced2396-bc8e-11e6-9e2e-3f0ee1c98e95.png)

**Comment by elonmallin** [comment]
Hey, yea it was `automatic` but I changed it to `fixed 192.168.1.1` as I said above and that didn't work. But tried the default fixed IP `fixed 8.8.8.8` and that actually works now, thanks!
```
C:\Users\atodw_000>docker run --rm --net=host appropriate/curl --head https://raw.githubusercontent.com/eclipse/che/master/cli.sh
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0 31247    0     0    0     0      0     HTTP/1.1 200 OK00:02 --:--:--     0
 0 --:--:--  0:00:02 --:--:--     0
Content-Security-Policy: default-src 'none'; style-src 'unsafe-inline'
Strict-Transport-Security: max-age=31536000
X-Content-Type-Options: nosniff
...
```

Maybe something about this can be added to the docs somewhere as troubleshooting or something?

**Comment by TylerJewell** [comment]
We can probably add a test in the new CLI to do a curl to a DNS entry, and if that fails try a ping to 8.8.8.8.  If that is successful, then we know that you have a Docker DNS resolver issue.

**Comment by elonmallin** [comment]
Okey sounds good. Now `che dir init` and `che dir up` works, so that's great.

However, if you don't mind, I created a new folder and did `che dir init` and edited the "Chefile" to add `workspace.runtime.image.location="node:6"` at line 3 and got this error:
```
WARNING: No kernel memory limit support
ERROR: ECLIPSE CHE(dir): Error: Error while parsing the file '/d/Projects/Tests/che-dir/Chefile' at line 3 and column 33. The error is :Cannot set property 'location' of undefined
```
I also have `CHE_VERSION=nightly`, `CHE_UTILITY_VERSION=nightly` and `CHE_DATA=/d/Projects...somefolder` set. And I ran `che update` already. The D drive is available to my containers in the docker settings too. So it seems like `image` is null for some reason?

**Comment by TylerJewell** [comment]
Let's not allow a single issue to drift off into different kinds of issues.  It makes it difficult to track.  For this issue, going to leave it open until we add some checks around the DNS vs. ping, and then we can close it out once we add in the right checks into the CLI around networking.

**Comment by TylerJewell** [comment]
@elonmallin - I am reviewing the code to see about adding this check, and wondering if this check is not needed with the new CLI.  In the CLI where you originally found this problem, we were using a curl command in order to connect to GitHub and to download files from GitHub.  In the new CLI, the only networking access that we would require is when we connect to dockerhub to download Docker images.  Which, in your particular configuratoin, actually worked.

We do not need appropriate/curl in the new CLI as the CLI is in a container and that container already has curl inside of it.  And even then, the curl commands that we do are only internal - it's for checking that Che is running locally.  So I am wondering if this check may not be necessary.

Once we get a new nightly with the latest changes, I will let you know about it and ask for a quick test, as I think this problem resolves itself in the new CLI.

Could you - as a preliminary test, rerun the new CLI syntax with the fixed DNS entry that you have.  Does it advance?

```
docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock 
-v /d/Projects/Tests/che-cli:/che eclipse/che-cli:nightly start
```

If it works with the fixed DNS and then fails without the fixed DNS, then there are some issues that we need to solve.  You should remove all of your docker images "docker rmi" before you test this because our CLI will use the cached images if possible.

**Comment by elonmallin** [comment]
Okey, sounds good. Please also explain how I should get the new cli later, if it's just a like that `eclipse/che-cli` image or still that bat and sh file but from another location or something.

Yea I wasn't sure since this is labeled as a question if we could just go on and on but that's not really helpful for keeping track of things so opened a new issue for the other thing here #3317.

**Comment by TylerJewell** [comment]
@elonmallin - Ok, added a quick fix for this into the new CLI.  It's in this PR:
https://github.com/eclipse/che/pull/3301

This will check to see if the DNS resolver is not working and then print out an error message and how to fix it.

**Comment by elonmallin** [comment]
Removed all my images and tested with fixed 8.8.8.8 and automatic. Fixed works and automatic gets your newly added INFO message:
```
PS D:\Projects\Tests\che-cli> docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock -v /d/Projects/Tests/che-
cli:/che -v /d/Projects/Tests/che-cli/data:/data eclipse/che-cli:nightly start
INFO: Welcome to Eclipse Che!
INFO:
INFO: We could not resolve DockerHub using DNS.
INFO: Either we cannot reach the Internet or Docker's DNS resolver needs a modification.
INFO:
INFO: You can:
INFO:   1. Modify Docker's DNS settings.
INFO:      a. Docker for Windows & Mac have GUIs for this.
INFO:      b. Typically setting DNS to 8.8.8.8 fixes resolver issues.
INFO:   2. Does your network require Docker to use a proxy?
INFO:      a. Docker for Windows & Mac have GUIs to set proxies.
INFO:   3. Verify that you have access to DockerHub.
INFO:      a. Try 'curl --head dockerhub.com'
```

**Comment by TylerJewell** [comment]
In the new version of the CLI, we inherit proxy configuration settings from your Docker daemon. So in Docker for Windows there is a location where you can set proxy or no_proxy information.

Glad to see the new check is working!
